draft-ietf-tls-esni.md

```diff
@@ -882,6 +893,14 @@ desired.
 Moreover, as noted in the introduction, SNI encryption is less useful
 without encryption of DNS queries in transit via DoH or DPRIVE mechanisms.
 
+## Optional Record Digests and Trial Decryption
+
+Supporting optional record digests and trial decryption opens oneself up to
+DoS attacks. Specifically, an adversary may send malicious ClientHello messages, i.e.,
+those which will not decrypt with any known ESNI key, in order to force
+decryption. Servers that support this feature should, for example, implement
+some form of rate limiting mechanism to limit the damage caused by such attacks.
+
```

Comment:
Looking at this, I wonder whether there are other mitigations than "rate limiting". Take for example the case of a server that has published only one key: trial decryption then is no more expensive that decryption directed by a specific checksum, which the adversary can learn from the DNS record. So an obvious mitigation in case of stress is to only trial one of the published keys, possibly the latest, or a randomly chosen one.

Maybe we should point out the directed DOS attack -- for example, an adversary could bomb a forwarding server with repeated connection request. But then, that's not much more potent than repeated non-ESNI connection requests.

Reply:
> I wonder whether there are other mitigations than "rate limiting".

There likely probably are, which is why I list this as only one example. :)